src/backend/distributed/planner/multi_planner.c

```diff
@@ -695,11 +717,19 @@ multi_relation_restriction_hook(PlannerInfo *root, RelOptInfo *relOptInfo, Index
 		return;
 	}
 
+	/*
+	 * Use a memory context that's guaranteed to live long enough, could be
+	 * called in a more shorted lived one (e.g. with GEQO).
+	 */
+	plannerRestrictionContext = CurrentPlannerRestrictionContext();
+	restrictionsMemoryContext = plannerRestrictionContext->memoryContext;
+	oldMemoryContext = MemoryContextSwitchTo(restrictionsMemoryContext);
+
 	distributedTable = IsDistributedTable(rte->relid);
 	localTable = !distributedTable;
 
-	restrictionContext = CurrentRelationRestrictionContext();
-	Assert(restrictionContext != NULL);
+	relationRestrictionContext = plannerRestrictionContext->relationRestrictionContext;
```

Comment:
This initialization might be better-placed on line 750, right above where it's first used. And I'm not sure this `Assert` serves any purpose either, it's either a crash here or a crash on line 751.